src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -92,21 +96,36 @@ function frule((_, ΔA), ::typeof(eigen!), A::StridedMatrix{T}; kwargs...) where
 end
 
 function rrule(::typeof(eigen), A::StridedMatrix{T}; kwargs...) where {T<:Union{Real,Complex}}
-    F = eigen(A; kwargs...)
+    # NOTE: this check for hermitian-ness occurs in `eigen!`. We here do it in the rrule for
+    # `eigen` so that this works for non-mutating AD
+    isherm = ishermitian(A)
+    if isherm
+        hermA, back_Hermitian = rrule(Hermitian, A, :U)
+        F, back_eigen = rrule(eigen, hermA; kwargs...)
+    else
+        F = eigen(A; kwargs...)
+    end
     function eigen_pullback(ΔF::Composite{<:Eigen})
-        λ, V = F.values, F.vectors
-        Δλ, ΔV = ΔF.values, ΔF.vectors
-        if ΔV isa AbstractZero
-            Δλ isa AbstractZero && return (NO_FIELDS, Δλ + ΔV)
-            ∂K = Diagonal(Δλ)
-            ∂A = V' \ ∂K * V'
+        if isherm
```

Comment:
Adding this check causes the rrule to become type-unstable. Not certain how to do this type-stably.

Reply:
I think I figured it out